stdlib/messaging/nats/src/main/java/org/ballerinalang/nats/basic/consumer/Detach.java

```diff
@@ -0,0 +1,84 @@
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.ballerinalang.nats.basic.consumer;
+
+import io.nats.client.Dispatcher;
+import org.ballerinalang.jvm.BallerinaErrors;
+import org.ballerinalang.jvm.scheduling.Strand;
+import org.ballerinalang.jvm.values.MapValue;
+import org.ballerinalang.jvm.values.ObjectValue;
+import org.ballerinalang.model.types.TypeKind;
+import org.ballerinalang.natives.annotations.BallerinaFunction;
+import org.ballerinalang.natives.annotations.Receiver;
+import org.ballerinalang.nats.Constants;
+import org.ballerinalang.nats.Utils;
+
+import java.io.PrintStream;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Unsubscribe the consumer from the subject.
+ *
+ * @since 1.0.4
+ */
+@BallerinaFunction(
+        orgName = Constants.ORG_NAME,
+        packageName = Constants.NATS,
+        functionName = "detach",
+        receiver = @Receiver(type = TypeKind.OBJECT,
+                structType = Constants.NATS_LISTENER,
+                structPackage = Constants.NATS_PACKAGE),
+        isPublic = true
+)
+public class Detach {
+    private static final PrintStream console;
+
+    public static Object detach(Strand strand, ObjectValue listener, ObjectValue service) {
+        @SuppressWarnings("unchecked")
+        List<ObjectValue> serviceList =
+                (List<ObjectValue>) ((ObjectValue) listener.get(Constants.CONNECTION_OBJ))
+                        .getNativeData(Constants.SERVICE_LIST);
+        MapValue<String, Object> subscriptionConfig = Utils.getSubscriptionConfig(service.getType()
+                .getAnnotation(Constants.NATS_PACKAGE, Constants.SUBSCRIPTION_CONFIG));
+        if (subscriptionConfig == null) {
+            return BallerinaErrors.createError(Constants.NATS_ERROR_CODE,
+                    "Error occurred while un-subscribing, Cannot find subscription configuration");
+        }
+        @SuppressWarnings("unchecked")
+        ConcurrentHashMap<String, Dispatcher> dispatcherList = (ConcurrentHashMap<String, Dispatcher>)
+                listener.getNativeData(Constants.DISPATCHER_LIST);
+        String subject = subscriptionConfig.getStringValue(Constants.SUBJECT);
+        Dispatcher dispatcher = dispatcherList.get(service.getType().getName());
+        try {
+            dispatcher.unsubscribe(subject);
+        } catch (IllegalArgumentException | IllegalStateException ex) {
+            return BallerinaErrors.createError(Constants.NATS_ERROR_CODE,
+                    "Error occurred while un-subscribing " + ex.getMessage());
+        }
+        console.println(Constants.NATS_CLIENT_UNSUBSCRIBED + subject);
```

Comment:
Just wondering whethe can log any client information like IP or something along with this message?

Reply:
None of the messaging connectors are logging that information anywhere right now.

Reply:
But right now this is a not useful information IMO.

Reply:
It is just to inform the user that it started to consume messages. Otherwise it doesn't show anything in the console, and the user will not know if the service started or not.

Reply:
This is more useful when attaching and detaching services.